sdk/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessEvaluationContextTest.java

```diff
@@ -0,0 +1,405 @@
+        ImmutableList.<AppliedPTransform<?, ?, ?>>of(
+            downstream.getProducingTransformInternal(), view.getProducingTransformInternal()));
+    valueToConsumers.put(downstream, ImmutableList.<AppliedPTransform<?, ?, ?>>of());
+    valueToConsumers.put(view, ImmutableList.<AppliedPTransform<?, ?, ?>>of());
+
+    Map<AppliedPTransform<?, ?, ?>, String> stepNames = new HashMap<>();
+    stepNames.put(created.getProducingTransformInternal(), "s1");
+    stepNames.put(downstream.getProducingTransformInternal(), "s2");
+    stepNames.put(view.getProducingTransformInternal(), "s3");
+
+    Collection<PCollectionView<?>> views = ImmutableList.<PCollectionView<?>>of(view);
+    context = InProcessEvaluationContext.create(
+            runner.getPipelineOptions(),
+            rootTransforms,
+            valueToConsumers,
+            stepNames,
+            views);
+  }
+
+  @Test
+  public void writeToViewWriterThenReadReads() {
+    PCollectionViewWriter<Integer, Iterable<Integer>> viewWriter =
+        context.createPCollectionViewWriter(
+            PCollection.<Iterable<Integer>>createPrimitiveOutputInternal(
+                p, WindowingStrategy.globalDefault(), IsBounded.BOUNDED),
+            view);
+    BoundedWindow window = new TestBoundedWindow(new Instant(1024L));
+    BoundedWindow second = new TestBoundedWindow(new Instant(899999L));
+    WindowedValue<Integer> firstValue =
+        WindowedValue.of(1, new Instant(1222), window, PaneInfo.ON_TIME_AND_ONLY_FIRING);
+    WindowedValue<Integer> secondValue =
+        WindowedValue.of(
+            2, new Instant(8766L), second, PaneInfo.createPane(true, false, Timing.ON_TIME, 0, 0));
+    Iterable<WindowedValue<Integer>> values = ImmutableList.of(firstValue, secondValue);
+    viewWriter.add(values);
+
+    SideInputReader reader =
+        context.createSideInputReader(ImmutableList.<PCollectionView<?>>of(view));
+    assertThat(reader.get(view, window), containsInAnyOrder(1));
+    assertThat(reader.get(view, second), containsInAnyOrder(2));
+
+    WindowedValue<Integer> overrittenSecondValue =
+        WindowedValue.of(
+            4444, new Instant(8677L), second, PaneInfo.createPane(false, true, Timing.LATE, 1, 1));
+    viewWriter.add(Collections.singleton(overrittenSecondValue));
+    assertThat(reader.get(view, second), containsInAnyOrder(4444));
+  }
+
+  @Test
+  public void getExecutionContextDifferentKeysIndependentState() {
+    InProcessExecutionContext fooContext =
+        context.getExecutionContext(created.getProducingTransformInternal(), "foo");
+
+    StateTag<Object, BagState<Integer>> intBag = StateTags.bag("myBag", VarIntCoder.of());
+
+    fooContext
+        .getOrCreateStepContext("s1", "s1", null)
+        .stateInternals()
+        .state(StateNamespaces.global(), intBag)
+        .add(1);
+
+    InProcessExecutionContext barContext =
+        context.getExecutionContext(created.getProducingTransformInternal(), "bar");
+    assertThat(
```

Comment:
may also be worth an assert not same as a simple "fail early and be super specific". Probably not necessary since you're testing that modifications to one don't show up in the other. Do you need to test that the modifications to state actually show up in the one that was modified?